activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -409,10 +409,11 @@ def construct_relation_for_association_calculations
       def apply_join_dependency(relation, join_dependency)
         relation = relation.except(:includes, :eager_load, :preload).joins!(join_dependency)
 
-        if using_limitable_reflections?(join_dependency.reflections)
+        if using_limitable_reflections?(join_dependency.reflections) ||
+            relation.group_values.any? || relation.having_clause.any?
```

Comment:
This doesn't read right to me: the existing condition is taking a shortcut if that shortcut is known to be safe, while the new alternatives are allowing additional paths when that condition has specifically declared it *not* safe to limit.